src/commands/serve.rs

```diff
@@ -0,0 +1,201 @@
+        &self.table
+    }
+
+    fn table_mut(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&self) -> &WasiCtx {
+        &self.ctx
+    }
+
+    fn ctx_mut(&mut self) -> &mut WasiCtx {
+        &mut self.ctx
+    }
+}
+
+impl WasiHttpView for Host {
+    fn table(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&mut self) -> &mut WasiHttpCtx {
+        &mut self.http
+    }
+}
+
+/// Runs a WebAssembly module
+#[derive(Parser)]
+#[structopt(name = "run")]
+pub struct ServeCommand {
+    #[clap(flatten)]
+    common: CommonOptions,
+
+    /// Socket address for the web server to bind to. Defaults to 0.0.0.0:8080.
+    #[clap(long = "addr", value_name = "SOCKADDR")]
+    addr: Option<std::net::SocketAddr>,
+
+    /// The WebAssembly component to run.
+    #[clap(value_name = "WASM", required = true)]
+    component: PathBuf,
+}
+
+impl ServeCommand {
+    fn addr(&self) -> std::net::SocketAddr {
+        self.addr.unwrap_or("0.0.0.0:8080".parse().unwrap())
+    }
+
+    /// Start a server to run the given wasi-http proxy component
+    pub fn execute(mut self) -> Result<()> {
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_time()
+            .enable_io()
+            .build()?;
+
+        runtime.block_on(async move {
+            tokio::select! {
+                _ = tokio::signal::ctrl_c() => {
+                    Ok::<_, anyhow::Error>(())
+                }
+
+                res = self.serve() => {
+                    res
+                }
+            }
+        })?;
+
+        Ok(())
+    }
+
+    fn add_to_linker(&self, linker: &mut Linker<Host>) -> Result<()> {
+        wasmtime_wasi_http::proxy::add_to_linker(linker)?;
+        Ok(())
+    }
+
+    async fn serve(&mut self) -> Result<()> {
+        use hyper::server::conn::http1;
+
+        let mut config = self.common.config(None)?;
+        config.wasm_component_model(true);
+        config.async_support(true);
+
+        let engine = Arc::new(Engine::new(&config)?);
+        let mut linker = Linker::new(&engine);
+
+        self.add_to_linker(&mut linker)?;
+
+        let component = Component::from_file(&engine, &self.component)?;
+
+        let instance = Arc::new(linker.instantiate_pre(&component)?);
+
+        let listener = tokio::net::TcpListener::bind(self.addr()).await?;
+
+        loop {
+            let (stream, _) = listener.accept().await?;
+            let engine = Arc::clone(&engine);
+            let instance = Arc::clone(&instance);
+            tokio::task::spawn(async move {
+                let handler = ProxyHandler::new(engine, instance);
+                if let Err(e) = http1::Builder::new()
+                    .keep_alive(true)
+                    .serve_connection(stream, handler)
+                    .await
+                {
+                    eprintln!("error: {e:?}");
+                }
+
+                Ok::<_, anyhow::Error>(())
```

Comment:
I think this can probably be dropped because the task doesn't return `Err(_)` anywhere